### PR TITLE
Improve deprecated external storage backend handling

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
@PVince81 you were right, the permissions model I thought up was stupid and complex. Ultimately it didn't even cover the one usecase it was supposed to, which is handling of deprecated backends.

I've reverted the permissions stuff and re-introduced 'visibilities', while extending them with the concept of 'deprecate to' properties on backends and auth mechanisms. SMB_OC and SFTP_Key deprecate to SMB and SFTP respectively. Deprecated backends are not shown in the various lists in the external storage GUI, but are still linked to the replacement backend when user mounting options are changed. E.g. if the admin deselects SMB for user mounting, it implicitly disables SMB_OC (but doesn't show it).

As it stands at the moment in master, the two deprecated backends aren't handled very well, appearing as 'SMB / CIFS with OC login [DEPRECATED]' or 'SFTP with secret key [DEPRECATED]'.

cc @icewind1991 @PVince81 @DeepDiver1975 

cc @jancborchardt for UX stuff

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
So we drop stuff in 9.0 which was introduced in 8.2?

Let's do this right away in the scope of 8.2

**Comment by Xenopathic** [comment]
@DeepDiver1975 Good idea, the permissions model was a crap idea. And this one works better.

**Comment by icewind1991** [comment]
:+1: looks good

**Comment by DeepDiver1975** [comment]
I'd like to hear @PVince81's opinion - THX

File: apps/files_external/service/backendservice.php [hunk 1]
```diff
@@ -31,17 +31,13 @@
...
-	/** Permission constants for PermissionsTrait */
-	const PERMISSION_NONE = 0;
-	const PERMISSION_MOUNT = 1;
-	const PERMISSION_CREATE = 2;
-	const PERMISSION_MODIFY = 4;
+	/** Visibility constants for VisibilityTrait */
+	const VISIBILITY_NONE = 0;
+	const VISIBILITY_PERSONAL = 1;
+	const VISIBILITY_ADMIN = 2;
+	//const VISIBILITY_ALIENS = 4;
```
Comment (PVince81) [inline apps/files_external/service/backendservice.php]: ancient aliens meme resurfaces - imgur

**Comment by PVince81** [comment]
What parts will need to be (regression-) retested ? I don't mind the change but hope it won't cause regressions.

**Comment by Xenopathic** [comment]
The visibility stuff was there before #18441, so that stuff has already been tested. We need to make sure that disabling user mount backends is still working, especially with the now deprecated backends. So, set up a SMB_OC or SFTP_Key storage, change to this branch, (make sure everything still works), disable SFTP and/or SMB for user mounting, make sure the SMB_OC and SFTP_Key backends are also disabled. It's the user mounting stuff that will be affected by this change.

**Comment by PVince81** [comment]
It didn't work for me:

1. Check out stable8.1
1. Create a user "user1"
1. Setup SFTP_Key with key as system-wide mount
1. Allow SFTP_Key and SMB_OC
1. Login as "user1"
1. Setup SFTP_Key as personal mount
1. Upgrade to this branch
1. See that both mounts still work
1. Login as admin
1. Uncheck the checkboxes "SFTP" and "SMB"
1. Check "Webdav" (to make sure it's not completely disabled)
1. Login as "user1"

At this point I can still see both mount points in the file list and personal page, and I can use them.

**Comment by Xenopathic** [comment]
@PVince81 Hmm, that's actually broken on master too :see_no_evil: :boom:

**Comment by PVince81** [comment]
Had a chat with @Xenopathic who helped to clarify the purpose of this PR.

:+1: this can be merged.

@Xenopathic can you have a look at the other unrelated issue that happens on master ? I seem to remember that it used to work in the past. Thanks.

**Comment by Xenopathic** [comment]
@PVince81 will do, looks like it's only master (phew), I need to put in a visibility check before storages are mounted

**Comment by PVince81** [comment]
Raised here https://github.com/owncloud/core/issues/19312

**Comment by LukasReschke** [comment]
Regression: https://github.com/owncloud/core/issues/19601

**Pull request merged**